### PR TITLE
Allow to setup compression modules for the training-only modules

### PR DESCRIPTION
**Pull request by ljaljushkin**
It's needed for elastic width in case of Aux branches, e.g. in InceptionV3.
We still don't need to compress these branches, but we need Mixin modules with pre_ops to propagate changes in the width.

Suggested approach - simple replace eval graph with train one and mark all training-only nodes to filter them in all previous algorithms. 

**Why is it draft?** 
There're some unresolved issues with statistics and setting compression. Looks like some compression is set for training only layers. Also wrapping training-only layers leads to some issues with LSTM.

### CODE REVIEW

**Review by vshampor — APPROVED** [review]

**Comment by ljaljushkin** [comment]
Jenkins please stop all builds

File: nncf/torch/graph/graph_builder.py [hunk 1]
```diff
@@ -34,27 +34,32 @@ def __init__(self, custom_forward_fn: Callable[[torch.nn.Module], Any]):
...
+    def convert(eval_dynamic_graph: DynamicGraph, train_dynamic_graph: DynamicGraph,
+                input_infos: List[ModelInputInfo] = None) -> PTNNCFGraph:
         layer_name_vs_node_counts = {}  # type: Dict[LayerName, int]
+        eval_graph_nodes = []
+        for dynamic_graph_node in eval_dynamic_graph.get_all_nodes():
+            layer_name = str(dynamic_graph_node.op_exec_context.op_address)
+            eval_graph_nodes.append(layer_name)
 
-        for dynamic_graph_node in dynamic_graph.get_all_nodes():
+        for dynamic_graph_node in train_dynamic_graph.get_all_nodes():
```
Comment (ljaljushkin) [inline nncf/torch/graph/graph_builder.py:54]: Simply use train dynamic graph as a base for insertion points.

File: nncf/torch/graph/graph_builder.py [hunk 2]
```diff
@@ -80,9 +85,10 @@ def convert(dynamic_graph: DynamicGraph, input_infos: List[ModelInputInfo] = Non
                                      ignored_algorithms=dynamic_graph_node.ignored_algorithms,
                                      is_in_iteration_scope=dynamic_graph_node.is_in_iteration_scope,
                                      is_integer_input=is_integer_input,
-                                     is_shared=is_shared)
+                                     is_shared=is_shared,
+                                     is_train_only_node=is_train_only_node)
```
Comment (ljaljushkin) [inline nncf/torch/graph/graph_builder.py:89]: Simply add attribute to nncf nodes

File: nncf/torch/pruning/filter_pruning/algo.py [hunk 3]
```diff
@@ -620,7 +620,7 @@ def _apply_binary_mask_to_module_weight_and_bias(module: torch.nn.Module,
             types_to_apply_mask.append('batch_norm')
 
         pruned_node_modules = list()
-        for node in graph.get_all_nodes():
+        for node in graph.get_all_eval_nodes():
```
Comment (ljaljushkin) [inline nncf/torch/pruning/filter_pruning/algo.py:623]: keep using eval nodes in all algorithms. compression for training-only nodes is required for NAS (width propagation)

File: tests/torch/data/reference_graphs/magnitude_sparsity/lstm_uni_seq.dot [hunk 4]
```diff
@@ -1,59 +1,52 @@
...
-"15 NNCF_RNN/StackedRNN[rnn_impl]/ModuleList[inners]/Recurrent[0]/LSTMCellForwardNNCF[cell]/__mul___2" -> "17 NNCF_RNN/StackedRNN[rnn_impl]/ModuleList[inners]/Recurrent[0]/cat_0";
-"15 NNCF_RNN/StackedRNN[rnn_impl]/ModuleList[inners]/Recurrent[0]/LSTMCellForwardNNCF[cell]/__mul___2" -> "20 NNCF_RNN/StackedRNN[rnn_impl]/cat_0";
-"17 NNCF_RNN/StackedRNN[rnn_impl]/ModuleList[inners]/Recurrent[0]/cat_0" -> "18 NNCF_RNN/StackedRNN[rnn_impl]/ModuleList[inners]/Recurrent[0]/view_0";
-"18 NNCF_RNN/StackedRNN[rnn_impl]/ModuleList[inners]/Recurrent[0]/view_0" -> "19 NNCF_RNN/StackedRNN[rnn_impl]/StackedRNNResetPoint/cat_0";
-"19 NNCF_RNN/StackedRNN[rnn_impl]/StackedRNNResetPoint/cat_0" -> "26 /nncf_model_output_2";
-"20 NNCF_RNN/StackedRNN[rnn_impl]/cat_0" -> "21 NNCF_RNN/StackedRNN[rnn_impl]/view_0";
-"21 NNCF_RNN/StackedRNN[rnn_impl]/view_0" -> "24 /nncf_model_output_0";
-"22 NNCF_RNN/StackedRNN[rnn_impl]/cat_1" -> "23 NNCF_RNN/StackedRNN[rnn_impl]/view_1";
-"23 NNCF_RNN/StackedRNN[rnn_impl]/view_1" -> "25 /nncf_model_output_1";
+"1 NNCF_RNN/StackedRNN[rnn_impl]/ModuleList[inners]/Recurrent[0]/LSTMCellForwardNNCF[cell]/NNCFLinear[input_linear]/linear_0" [id=1, type=linear];
```
Comment (ljaljushkin) [inline tests/torch/data/reference_graphs/magnitude_sparsity/lstm_uni_seq.dot:3]: TODO: fix missing `UpdateWeight` operations!!

**Pull request closed without merging**